Spark: Encapsulate parquet objects for Comet

parthchandra opened the pull request:
Fixes issues in https://github.com/apache/iceberg/pull/13378. This makes all the remaining changes to Iceberg to address issues with Comet/Parquet shading. The corresponding changes to Comet are already merged into Comet 0.10.1 
Additionally, the dependency of the iceberg-parquet module has been removed because it created an iceberg-flink dependency on comet-spark. The code is changed to use a reflection based CometBridge class to access the come (Parquet) `FileReader`.
This PR also removes the commit to enable Comet execution (moving it to a follow up PR: cc @hsiang-c).

This PR makes the changes for Spark 3.4, 3.5, and 4.0

parthchandra commented:
@huaxingao, @anuragmantri  please take a look.

hsiang-c commented on parquet/src/main/java/org/apache/iceberg/parquet/CometVectorizedParquetReader.java:
```diff
@@ -0,0 +1,255 @@
+  private ReadConf init() {
+    if (conf == null) {
+      ReadConf readConf =
+          new ReadConf(
```
This constructor will create a `ParquetFileReader` internally, do you think we could use Comet's `FileReader`?

parthchandra replied:
I created a Comet specific `CometReadConf` that does not have a `ParquetFileReader` but uses the Comet `FileReader` instead and changed `CometVectorizedParquetReader` to reuse the FileReader created in `CometReadConf`.
Unfortunately, `CometVectorizedParquetReader` needs the file schema and `RowGroups`  from `CometReadConf` and using Comet's `FileReader` to get this information in Iceberg violates the requirement to not use any Comet API that has a Parquet class in the signature (doing so breaks the implementation when Parquet is  shaded by Iceberg). Also, the methods to get schema and row groups are also (rightly) made package-private in Comet so the change won't even compile. 
I think for the moment this will have to do. We will revisit this integration once again as we move to a more native implementation in Comet.

hsiang-c commented on parquet/src/main/java/org/apache/iceberg/parquet/CometVectorizedParquetReader.java:
```diff
@@ -0,0 +1,255 @@
+      }
+      RowGroupReader pages;
+      try {
+        pages = cometReader.readNextRowGroup();
```
`readNextRowGroup` might return `null` and `setRowGroupInfo` will make use of `pages`.

parthchandra replied:
In this case, `advance` is called only by `next` which is implementing the `iterator` interface and which explicitly checks for `hasNext` so `readNextRowGroup` always returns a not null value.
This could be done better, but I'll leave this as it is because - 
 i) it is not really clear what the behavior will be if `advance` silently does not do anything. 
 ii) the code (and behavior) will be consistent with the implementation in `VectorizedParquetReader`

parthchandra commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -1352,16 +1358,35 @@ public <D> CloseableIterable<D> build() {
-              caseSensitive,
-              maxRecordsPerBatch);
+          if (isComet) {
+            LOG.info("Comet enabled");
```
changed

shangxinli commented on parquet/src/main/java/org/apache/iceberg/parquet/CometTypeUtils.java:
```diff
@@ -0,0 +1,260 @@
+          throw new IllegalArgumentException(
+              "Missing required parameters for DecimalLogicalTypeAnnotation: " + params);
+        }
+        int scale = Integer.parseInt(params.get("scale"));
```
Should we add validation for the parameter format? for example, 
If params.get("scale") returns null we will get NumberFormatException. Another case is non-numeric string.

parthchandra replied:
It makes the code very verbose, but I've added the additional validation.

shangxinli commented on parquet/src/main/java/org/apache/iceberg/parquet/CometVectorizedParquetReader.java:
```diff
@@ -0,0 +1,255 @@
+          specs.add(spec);
+        }
+
+        fileReader.setRequestedSchemaFromSpecs(specs);
```
If setRequestedSchemaFromSpecs fails, the FileReader will be closed?

parthchandra commented:
@shangxinli thank you for taking a look. I found that there are more changes to get around parquet shading that need to be addressed; let me address your comments along with that.

huaxingao commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -1422,16 +1428,35 @@ public <D> CloseableIterable<D> build() {
-              maxRecordsPerBatch);
+          if (isComet) {
+            LOG.info("Comet vectorized reader enabled");
+            return new CometVectorizedParquetReader<>(
```
@pvary I’ve seen your work on the reader/writer API proposal, but I haven’t kept up with all the details. We’re adding a `CometVectorizedParquetReader`, does this align with your proposal?

parthchandra commented:
@huaxingao thanks for bringing up the Spark 3.4 discussion in the mailing list. At this point, we could still move ahead with this PR as it is and if 3.4 is dropped, then the additional code here will also get dropped. Otherwise if 3.4 is dropped before we merge this then I will have to rebase and will drop 3.4 from here as well. (In that case Comet will have to provide instructions on manually patching and building Iceberg).
Meanwhile if there are any more items to address, then I will be happy to take care of them.

huaxingao commented on parquet/src/main/java/org/apache/iceberg/parquet/CometVectorizedParquetReader.java:
```diff
@@ -0,0 +1,287 @@
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.schema.MessageType;
+
+public class CometVectorizedParquetReader<T> extends CloseableGroup
```
Thanks @parthchandra for the detailed explanation. I’d prefer we keep this in the Spark module for now. If we later have a concrete need to use Comet’s Parquet reader with Flink, we can revisit and refactor accordingly.

parthchandra commented:
@huaxingao PTAL. I've moved _all_ Comet classes out of `iceberg-parquet`. The code now uses a Factory method that loads a vectorized parquet reader via the Java `ServiceLoader` interface if a custom reader is configured. Comet now provides this custom reader for each Spark version supported.
This approach also eliminates the 'bridge' classes.
The only downside is that the `CometVectorizedParquetReader` and the `CometTypeUtils` are duplicated. Also, Flink will no longer automatically get Comet support. (But can be configured to use the Comet reader if the jars are in the path).

Note: 
  I added `VectorizedParquetReaderFactory.java` in the `iceberg-parquet` module. This interface defines two methods: `name()` and `createReader(...)` for pluggable vectorized reader implementations. 
  Concrete implementations are in the Spark modules.

huaxingao commented on build.gradle:
```diff
@@ -845,6 +845,7 @@ project(':iceberg-orc') {
 }
 
 project(':iceberg-parquet') {
+
```
nit: unnecessary change

parthchandra commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -1352,6 +1367,32 @@ public <D> CloseableIterable<D> build() {
 
         if (batchedReaderFunc != null) {
+          // Try to load custom vectorized reader factory from properties
+          String readerName = properties.get("read.parquet.vectorized-reader.factory");
```
Done

parthchandra commented on spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/parquet/CometTypeUtils.java:
```diff
@@ -0,0 +1,260 @@
+  }
+
+  private static LogicalTypeAnnotation reconstructLogicalType(
+      String logicalTypeName, java.util.Map<String, String> params) {
```
Done

pvary commented on parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java:
```diff
@@ -60,7 +60,7 @@ class ReadConf<T> {
 
   @SuppressWarnings("unchecked")
-  ReadConf(
+  public ReadConf(
```
do we need this public?

parthchandra replied:
Yes we do. The Comet reader is now in a different module and needs to access the read conf for read information. I could make it protected and then derive a class explicitly meant for custom readers, but that seems overkill.

parthchandra replied:
I would rather not rename an existing class. I feel it will reduce confusion on the scope of the changes.

pvary commented on parquet/src/main/java/org/apache/iceberg/parquet/VectorizedParquetReaderFactory.java:
```diff
@@ -0,0 +1,90 @@
+   * @param <T> the type of records returned by the reader
+   * @return a closeable iterable of records
+   */
+  <T> CloseableIterable<T> createReader(
```
Or we could revisit adding one more parameter to the `FormatModelRegistry.readBuilder` so we don't have to hack with the different return class.

manuzhang commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java:
```diff
@@ -30,6 +30,15 @@ private SparkSQLProperties() {}
+  // Controls the fully qualified class name of the vectorized Parquet reader factory
+  public static final String PARQUET_VECTORIZED_READER_FACTORY =
+      "spark.sql.iceberg.parquet.vectorized-reader.factory";
+
```
nit: no space between related configs.

parthchandra replied:
fixed

parthchandra commented:
> What about something like this: [pvary@4b6f7a5](https://github.com/pvary/iceberg/commit/4b6f7a5f577afe0d5f4742e4602cc34c14c0115d)
> This builds above the proposed File Format API: #12774, #12298

@pvary I'm not sure what you're asking here. Would you like this PR to be changed to build on top of the proposed API?

parthchandra commented on spark/v4.0/build.gradle:
```diff
@@ -272,6 +272,8 @@ project(":iceberg-spark:iceberg-spark-runtime-${sparkMajorVersion}_${scalaVersio
     integrationImplementation project(path: ':iceberg-hive-metastore', configuration: 'testArtifacts')
     integrationImplementation project(path: ":iceberg-spark:iceberg-spark-${sparkMajorVersion}_${scalaVersion}", configuration: 'testArtifacts')
     integrationImplementation project(path: ":iceberg-spark:iceberg-spark-extensions-${sparkMajorVersion}_${scalaVersion}", configuration: 'testArtifacts')
+    integrationImplementation "org.apache.datafusion:comet-spark-spark3.5_2.13:${libs.versions.comet.get()}"
```
Corrected (This is a new line so rebasing didn't address this)

manuzhang commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -1442,6 +1471,29 @@ public <D> CloseableIterable<D> build() {
 
         if (batchedReaderFunc != null) {
+          // Try to load custom vectorized reader factory from properties
+          String readerName = properties.get(VECTORIZED_READER_FACTORY);
```
isn't this `factoryName`?

manuzhang commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkBatch.java:
```diff
@@ -137,10 +137,24 @@ public PartitionReaderFactory createReaderFactory() {
+    String factoryClassName = readConf.parquetVectorizedReaderFactory();
+
+    // If no explicit factory is set and reader type is COMET, use the default Comet factory
+    if (factoryClassName == null && readerType == ParquetReaderType.COMET) {
```
This can be handled inside `readConf.parquetVectorizedReaderFactory()` and moved to right before the check `if (factoryClassName != null) {`.

manuzhang commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/parquet/CometVectorizedParquetReader.java:
```diff
@@ -0,0 +1,352 @@
+  private ReadConf conf = null;
+
+  private ReadConf init() {
+    if (conf == null) {
```
Is this thread safe?

parthchandra replied:
I don't think it needs to be. This is pretty much what is done in VectorizedParquetReader

parthchandra commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -1537,6 +1589,27 @@ public <D> CloseableIterable<D> build() {
+      Class<?> factoryClass = Class.forName(className);
+      if (!VectorizedParquetReaderFactory.class.isAssignableFrom(factoryClass)) {
+        LOG.warn("Class {} does not implement VectorizedParquetReaderFactory interface", className);
+        return null;
```
We lose a little bit of helpful text in the error, but made the change.

parthchandra commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/parquet/CometVectorizedParquetReader.java:
```diff
@@ -0,0 +1,352 @@
+              reuseContainers,
+              caseSensitive,
+              batchSize);
+      this.conf = readConf.copy();
```
TBH, I don't know myself. Based it on what is being done in VectorizedParquetReader

manuzhang commented:
> The proposal to remove 3.4 has already been approved so we are racing against the clock here.

@parthchandra Targeting multiple Spark versions in one PR doesn't make it faster. Usually when a PR targets Spark 4.0 is merged, the back-port PR can be reviewed and merged quickly if the changes are identical. On the other hand, reviewers can get distracted if a PR targets multiple Spark versions.

anuragmantri commented:
The last [PR](https://github.com/apache/iceberg/pull/15328) for the file format API changes in Spark is now merged. @parthchandra would you like to update this PR?
